sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -475,6 +475,8 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
       case If(TrueLiteral, trueValue, _) => trueValue
       case If(FalseLiteral, _, falseValue) => falseValue
       case If(Literal(null, _), _, falseValue) => falseValue
+      case If(cond, TrueLiteral, FalseLiteral) if cond.deterministic => cond
```

Comment:
The `cond.deterministic` check does not seem to be necessary - it will be evaluated the same number of times in either case.

Reply:
We had similar discussions before: https://github.com/apache/spark/pull/21848#pullrequestreview-139604857

Reply:
It is a slightly different situation.
In the one you linked, the condition will not be evaluated after the optimization.
Here the condition will be evaluated the same amount of times with and without the optimizer - meaning this optimization is okay for even non-deterministic expressions.

Reply:
In this case, I agree with @tanelk . The following two will be the same.
```scala
If(GreaterThan(Rand(0), UnresolvedAttribute("a")), TrueLiteral, FalseLiteral)
GreaterThan(Rand(0), UnresolvedAttribute("a"))
```

Reply:
cc @maropu

Reply:
Agree with @tanelk here. `cond` is actually put in same relative position so it should be fine not to check `deterministic`.

Reply:
+1 on the @tanelk opinion, too.

Reply:
+1